### PR TITLE
HDDS-3134. Debug Tool that gives chunk location information given a key.

### PR DESCRIPTION
**Pull request by sadanand48**
## What changes were proposed in this pull request?
Debug tool to display chunk information given a key.Apart from Chunks ,it also displays information about the container on which the chunks are present and the pipeline info

**`Non-Verbose output:`**
Screenshot 2020-04-02 at 4 40 20 PM

**`Verbose Output:`**
Screenshot 2020-04-02 at 4 41 46 PM

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-3134

## How was this patch tested?
Included robot tests in ozone-shell.robot

### CODE REVIEW

**Comment by avijayanhwx** [comment]
@sadanand48 Please fix the failing CI issues in this patch.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueContainerData.java [hunk 1]
```diff
@@ -20,6 +20,8 @@
 
 import com.google.common.base.Preconditions;
 import com.google.common.collect.Lists;
+
+import java.awt.*;
```
Comment (bshashikant) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueContainerData.java:24]: plz Fix the imports

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/OzoneAdmin.java [hunk 2]
```diff
@@ -1,4 +1,4 @@
-/**
```
Comment (bshashikant) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/OzoneAdmin.java:1]: Unrelated change.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ChunkKeyHandler.java [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+      // localID pertaining to a given key
+      List<OzoneKeyLocation> keyLocationList = key.getOzoneKeyLocations();
+      for (OzoneKeyLocation keyLocation:keyLocationList) {
+        ContainerChunkInfo containerChunkInfoVerbose = new
+                ContainerChunkInfo();
+        ContainerChunkInfo containerChunkInfo = new ContainerChunkInfo();
+        chunkPaths.clear();
+        chunkDetailsList.clear();
+        long containerId = keyLocation.getContainerID();
+        BlockID blockID = new BlockID(containerId,
```
Comment (bshashikant) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ChunkKeyHandler.java:99]: we should use keyLocation.getBlockId() here

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ChunkKeyHandler.java [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+                keyLocation.getLocalID());
+        try (ScmClient scmClient = containerOperationClient) {
```
Comment (bshashikant) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ChunkKeyHandler.java:101]: Not required..Can you plz remove it?

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ChunkKeyHandler.java [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+                    + chunkInfo.getChunkName());
+          }
+          containerChunkInfoVerbose.setContainerPath(
+                  containerData.getContainerPath());
+          containerChunkInfoVerbose.setDataNodeList(
+                  container.getPipeline().getNodes());
+          containerChunkInfoVerbose.setPipeline(container.getPipeline());
+          containerChunkInfoVerbose.setChunkInfos(chunkDetailsList);
+          containerChunkInfo.setChunks(chunkPaths);
+          List<ChunkDataNodeDetails> chunkDataNodeDetails = new
```
Comment (bshashikant) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ChunkKeyHandler.java:138]: Add blockId to containerChunkInfo as well.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueContainerData.java [hunk 6]
```diff
@@ -24,7 +24,7 @@
 
 import org.apache.hadoop.hdds.protocol.datanode.proto.ContainerProtos;
 import org.apache.hadoop.hdds.protocol.datanode.proto.ContainerProtos
-    .ContainerDataProto;
```
Comment (bshashikant) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueContainerData.java:27]: unintended change

File: hadoop-ozone/dist/src/main/smoketest/basic/ozone-shell.robot [hunk 7]
```diff
@@ -60,7 +60,7 @@ Test ozone shell
                     Should contain      ${result}       creationTime
     ${result} =     Execute             ozone sh volume list | jq -r '. | select(.name=="${volume}")'
                     Should contain      ${result}       creationTime
-# TODO: Disable updating the owner, acls should be used to give access to other user.
```
Comment (bshashikant) [inline hadoop-ozone/dist/src/main/smoketest/basic/ozone-shell.robot:63]: unintended change

File: hadoop-ozone/dist/src/main/smoketest/basic/ozone-shell.robot [hunk 8]
```diff
@@ -116,13 +117,16 @@ Test key handling
...
     ${result} =     Execute             ozone sh key info ${protocol}${server}/${volume}/bb1/key1 | jq -r '. | select(.name=="key1")'
                     Should contain      ${result}       creationTime
+    ${result} =     Execute             ozone debug chunkinfo ${protocol}${server}/${volume}/bb1/key1 | jq  '.[]'
+                    Should contain      ${result}       chunks
     ${result} =     Execute             ozone sh key list ${protocol}${server}/${volume}/bb1 | jq -r '. | select(.name=="key1") | .name'
                     Should Be Equal     ${result}       key1
                     Execute             ozone sh key rename ${protocol}${server}/${volume}/bb1 key1 key2
     ${result} =     Execute             ozone sh key list ${protocol}${server}/${volume}/bb1 | jq -r '.name'
                     Should Be Equal     ${result}       key2
                     Execute             ozone sh key delete ${protocol}${server}/${volume}/bb1/key2
```
Comment (bshashikant) [inline hadoop-ozone/dist/src/main/smoketest/basic/ozone-shell.robot:127]: unwanted empty line

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ChunkKeyHandler.java (hunk 3 above)
Comment (bshashikant) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ChunkKeyHandler.java:98]: Can we change the code here to directly use keyLocation.getBlockID()

**Comment by sadanand48** [comment]
> The changes look good to me with some minor comments to be addressed.

**Comment by sadanand48** [comment]
> > The changes look good to me with some minor comments to be addressed.



> > The changes look good to me with some minor comments to be addressed.
